bcache-tools: fix udev patch to use bash from explicit path

pimeys opened the pull request:
## Things done

This fixes the issue in:

https://github.com/NixOS/nixpkgs/issues/349975

The bcache udev rule was patched to run `/bin/sh`, but at some point there was a check introduced which checked if all the binaries called in the udev rules actually exist. This one failed the check, and resulted a broken build. What is now fixed here is we explicitly set the path to the system bash, and the udev check is now successful.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pimeys committed:
bcache-tools: fix udev patch to use bash from explicit path

pimeys commented:
This touches your patch @simonhollingshead

pimeys commented:
And pinging the maintainer @bjornfor

bjornfor reviewed (commented):
> bcache-tools: fix udev patch to use bash from explicit path 

Can you summarize the problem and the fix in the commit message? Has this been broken for a long time? Which commit broke it?

Getting this info into the commit message isn't a hard requirement, but it does help making more sense of why changes are made in the repo.

simonhollingshead commented:
> This touches your patch @simonhollingshead

I don't agree that it touches 'my' patch.

That line with "/bin/sh" in it was introduced in 2014 (https://github.com/NixOS/nixpkgs/commit/92f94e901a4c750678c27cb5d0bdfd25706d483e).  I updated the package to a newer upstream version which changed other lines elsewhere in the file (removing a kmod load, removing insertion of a comment, dropping the custom bcache-register script).

I am unsure if you've misread the diff of my change because github highlighting for diff files shows green not only for lines inserted by the commit, but also for all '+' lines in a diff file.

As such, I have no opinion on this change and agree that the maintainer is the correct person to ping.

pimeys commented:
> > bcache-tools: fix udev patch to use bash from explicit path
> 
> Can you summarize the problem and the fix in the commit message? Has this been broken for a long time? Which commit broke it?
> 
> Getting this info into the commit message isn't a hard requirement, but it does help making more sense of why changes are made in the repo.

So as much as I know about this issue, I summarized it in here:

https://github.com/NixOS/nixpkgs/issues/349975

Switching to the new configuration is broken in rev `4c2fcb090b1f3e5b47eaa7bd33913b574a11e0a0` and was still working in rev `5633bcff0c6162b9e4b5f1264264611e950c8ec7`. This patch has _not_ changed between those commits, but something else in the system introduced a check which checks that all binaries are available in the udev rules. I tried to ask about this in the NixOS Matrix channel, but I didn't find anybody who could tell me what this change was.

@K900 seemed to agree using an explicit path here is a good idea.

bjornfor commented:
> Switching to the new configuration is broken in rev 4c2fcb090b1f3e5b47eaa7bd33913b574a11e0a0 and was still working in rev 5633bcff0c6162b9e4b5f1264264611e950c8ec7. This patch has not changed between those commits, but something else in the system introduced a check which checks that all binaries are available in the udev rules. I tried to ask about this in the NixOS Matrix channel, but I didn't find anybody who could tell me what this change was.

`git bisect` should tell you which commit broke your setup.

pimeys commented:
> > Switching to the new configuration is broken in rev [4c2fcb0](https://github.com/NixOS/nixpkgs/commit/4c2fcb090b1f3e5b47eaa7bd33913b574a11e0a0) and was still working in rev [5633bcf](https://github.com/NixOS/nixpkgs/commit/5633bcff0c6162b9e4b5f1264264611e950c8ec7). This patch has not changed between those commits, but something else in the system introduced a check which checks that all binaries are available in the udev rules. I tried to ask about this in the NixOS Matrix channel, but I didn't find anybody who could tell me what this change was.
> 
> `git bisect` should tell you which commit broke your setup.

But wouldn't this then be to try to build the whole system somehow with a different revision of nixpkgs? And what package this even is which breaks? It's somewhere in building `etc` or `udev-rules`, but how to test this directly from nixpkgs so I can utilize git-bisect?

pimeys commented:
The error is thrown from here in the current main, you can't build a nixos configuration with the latest unstable and main:

https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/hardware/udev.nix#L102

But this change is pretty old, something more recent just got rid of `/bin/sh` from the point it checks these udev rules.

bjornfor commented:
> > `git bisect` should tell you which commit broke your setup.
> 
> But wouldn't this then be to try to build the whole system somehow with a different revision of nixpkgs?

Yes.

> And what package this even is which breaks? It's somewhere in building `etc` or `udev-rules`, but how to test this directly from nixpkgs so I can utilize git-bisect?

It depends on how you build/deploy your system(s), but here's one way to do it: `nixos-rebuild build -I nixpkgs=/path/to/local/nixpkgs`

pimeys commented:
> > > `git bisect` should tell you which commit broke your setup.
> > 
> > 
> > But wouldn't this then be to try to build the whole system somehow with a different revision of nixpkgs?
> 
> Yes.
> 
> > And what package this even is which breaks? It's somewhere in building `etc` or `udev-rules`, but how to test this directly from nixpkgs so I can utilize git-bisect?
> 
> It depends on how you build/deploy your system(s), but here's one way to do it: `nixos-rebuild build -I nixpkgs=/path/to/local/nixpkgs`

Interestingly this issue _does not happen_ anymore when I was once able to build it with my patch. Like, there is some state somewhere which is not getting built anymore, and I can again build my system against nixos-unstable bcache-tools.

What seems to be the common agreement in NixOS matrix channel is that the udev rules should always point to the binaries they need explicitly, but nobody still knows how this regression happened with bcache-tools.

pimeys commented:
> @pimeys: I already pointed you to the cause in the issue that you opened.
> 
> The problem is https://git.sr.ht/~pimeys/nixos/tree/bff8a0c999e581f6f46ddb78151f2fd4e1a2bd80/item/core/default.nix#L82
> You add a sandbox path which will bind-mount `${pkgs.bash}/bin/sh` over `/bin/sh` in the sandbox. However, `${pkgs.bash}/bin/sh` is a symlink to `bash` in the same directory, and so because of the bind mount, this will resolve to `/bin/bash` in the sandbox, which doesn't exist.
> 
> This is why the check for the binaries fails for you, but doesn't for most people, because by default nix creates a `/bin/sh` in the sandbox that actually works.
> 
> So, I'm not against merging this PR, it's probably better in general to have fully-qualified paths in our udev rules that don't depend on the presence of `/bin/sh` in nixos (and in the sandbox). But your setup is fundamentally broken and you are likely to run into more issues like this down the line.
> 
> You could add another sandbox path to also create `/bin/bash` in the sandbox, which would make things work again. But I don't actually see why you configure your build sandbox like this to begin with, since nix already creates `/bin/sh` in the sandbox by default.

Interestingly I did remove that earlier and it was still failing. Like I said there is some cache that kept on building the same derivation, which was only fixed when I changed the derivation of bcache-tools.

https://git.sr.ht/~pimeys/nixos/commit/02583de0060e4a879bdaf8b5156a5b96e823b493

Also interestingly this has been working for years for me, and the reason I added this in 2021 is unclear today. Oh well, maybe we merge this and I also remove this hack from my config.

r-vdp commented:
> Interestingly I did remove that earlier and it was still failing. Like I said there is some cache that kept on building the same derivation, which was only fixed when I changed the derivation of bcache-tools.

When changing that config, you'd need to do a successful rebuild before it takes effect, since you actually need the modified config file in etc and you need to have restarted the nix deamon.
If you just change the value, you'll still be building with the previous config with a broken sandbox.
Maybe that's why it didn't seem to work?

> https://git.sr.ht/~pimeys/nixos/commit/02583de0060e4a879bdaf8b5156a5b96e823b493
> 
> Also interestingly this has been working for years for me, and the reason I added this in 2021 is unclear today. Oh well, maybe we merge this and I also remove this hack from my config.

Yeah, something must've changed recently. I'm actually surprised that you didn't encounter any other breakage in all that time. It's a good thing actually, it means that few things actually rely on `/bin/sh` being there.

r-vdp commented:
@bjornfor since the actual issue was due to a misconfiguration on @pimeys end, I don't think that there's much more to include in the commit message.
However, I still think that it's the right thing to do to have a fully-qualified path in the udev rule instead of relying on the presence of `/bin/sh`, so I'd still like to get this change in.

Do you agree?

bjornfor commented:
> Do you agree?

Yes.
